### ISSUE TITLE
HUD Support

### ISSUE THREAD

**Issue by rilian-la-te** [rilian-la-te/vala-panel-appmenu#67]
Enhancement as HUD support.
Planned after all external modules are integrated and dbusmenu punished.

**Comment by rilian-la-te** [comment]
@flexiondotorg How hud must work? Is it must search by all programs' menus or current menu only?

**Comment by flexiondotorg** [comment]
The menu of the currently active application only.

**Comment by rilian-la-te** [comment]
thanks.

**Comment by flexiondotorg** [comment]
Are you planning to add HUD support via vala-panel-appmenu or a separate applet/process? I'd like to be able to use the HUD without having to have Global Menu activated.

**Comment by rilian-la-te** [comment]
No separate applet:-) but I will allow to use it without global menu (but I
does not understand HUD importance)

**Comment by flexiondotorg** [comment]
OK, great :-) So long as HUD can function with or without Global Menu that will be great.

The HUD is really useful for quickly searching for and triggering menu entries without having your fingers leave the keyboard. This is most valuable in applications with extensive menu structures that are often slow to traverse and find the correct entry using a mouse. For example finding effects in Audacity or filters in GIMP is much quicker using the HUD.

Just tap Alt and type to filter the list of menus that match your search text then hit enter (or cursor down and couple and hit enter) and done :-)

**Comment by rilian-la-te** [comment]
So, write your feature requrements.
I does not like HUD and does not want to use it. And I want it to work
without X.

**Comment by snippins** [comment]
Just want to ask: I current use a modified version of i3-hud-menu (but use rofi instead of dmenu) to make a HUD. What i3-hud-menu did is get the menu entries through the org.gtk.Menus dbus interface. 

At first I notice that some application like Firefox or Visual Studio Code I could not get all the menu-entries to show in the HUD. However If I click the a menu (I use xfce4-panel) the HUD able to see the content of the that menu in the HUD. So how could I modifiy the plugin to expose all menu entries for me? I think you would need to do this if you want to the HUD anyway, right?

**Comment by rilian-la-te** [comment]
AFAIK, i3-hud-menu reads MenuModels only. But there are another protocol for global menus - DBusMenu. And VSCode use it, not MenuModels. Exposing MenuModels from plugin is side effect.

**Comment by snippins** [comment]
Sorry I was so sure vscode use MenuModel and did not mention that i3-hud-menu do also read DBusMenu. It does have a service that implement the Registra interface. However since vala plugin can be considered as implementation of that interface we do not need to run that service for i3-hud-menu to function. 

The hud does not show all commands is due to the way vala-appmenu is implemented actually. For example if I enable the service implement by KDE itself (Background Service->Application menu daemon) the HUD is able to show all commands from visual studio code. Thus if your plugin to expose it fully like KDE did the HUD should work. 

Also (as expected) enable the KDE daemon and use vala appmenu at the same time cause some weird bugs (due to race condition I think).

I actually do not think expose all the commands is how vala-appmenu need to work, I just want to have some directions to modify it myself so I can use my i3-hud-menu. Thanks.

**Comment by rilian-la-te** [comment]
1. Please, write bug about unsupported commands (I want to know about it, because I have implemented all Registrar AFAIK. 
2. What bugs you have, if you use vala-panel-appmenu with KDE Damon? 
3. What commands are not shown?

**Comment by snippins** [comment]
I have filed a bug #89. Thank you.

**Comment by snippins** [comment]
@rilian-la-te Just link my HUD implementation in python [here](https://gist.github.com/snippins/ee943f2b25db555ef12107f7cee20241) in case you need it.

**Comment by flexiondotorg** [comment]
@snippins I create `mate-hud` which you can find here:

  * https://bitbucket.org/ubuntu-mate/mate-hud

Are you interested in helping getting that project fixed up and working correctly? If so, please file issues and pull requests over there :-)

Despite what HUD capability is added to vala-panel-appmenu it would be good to get `mate-hud` fixed as best as possible.

**Comment by snippins** [comment]
@flexiondotorg I don't use mate so it is a little tricky to test. Also I want the HUD to be platform independent. However it is basically a one-file project I think it still very doable though. Maybe only after everything got fixed/stable on my side I will do a pull request for you. Also feel free to merge the code your self if you want. I copy a portion of you code actually :D.

**Comment by flexiondotorg** [comment]
Despite the name `mate-hud` is desktop agnostic, but favours GTK for automatic theme integration. Any pull requests welcomed and I'll look at your code more carefully tomorrow ☺️

**Comment by snippins** [comment]
@flexiondotorg I see. Anyway I just update the [HUD](https://gist.github.com/snippins/ee943f2b25db555ef12107f7cee20241) again, If possible you help to test if is stable for you, especially gtk apps (to use with unity-gtk-module + vala-panel-appmenu OR i3-hud-service.py). If you have any question about the code you can ask there also, thanks.

**Comment by flexiondotorg** [comment]
@rilian-la-te After having tested `mate-hud` when the vala-panel-appmenu registrar is available, they integrate very well. So I think my requirement for HUD support is now simply this: 

  * Expose a gsettings schema to have your registrar alway running regardless if the global menu applet is active. I can then simplify `mate-hud` to simply be a client of your registrar.

@snippins I'm going to try and find the time to merge your changes to `mate-hud` today.

**Comment by flexiondotorg** [comment]
@snippins I've merged the changes, I'd really appreciate if you could do some quick tests with `mate-hud.py` (just the client) in-conjunction with vala-panel-appmenu and confirm it works as you'd expect.

  * https://bitbucket.org/ubuntu-mate/mate-hud/commits/bb3e37edd550b67faab85a3a0f85cf8db04bf991?at=master

**Comment by flexiondotorg** [comment]
@snippins I've fixed up some bit and pieces and cleaned up. `mate-hud` is working with GTK3, Qt4, Qt5, Electron, Firefox, Thunderbird and LibreOffice. However, not working with GTK2 which is probably because I don't see any GTK2 applications having a global menu.

**Comment by snippins** [comment]
@flexiondotorg Could be the unity-gtk-module not loaded when logging in. It will work if that fixed. I am on KDE Neon and gtk2 apps like xfce4-terminal and sublime text have their HUDs work correctly.

**Comment by flexiondotorg** [comment]
@snippins Well the GTK3 applications works and `unity-gtk2-module` and `unity-gtk3-module`.

**Comment by flexiondotorg** [comment]
@snippins GTK2 is now working. I had the old TopMenu modules installed and exported which was preventing GTK2 applications from working.

**Comment by rilian-la-te** [comment]
why APPMENU_DISPLAY_BOTH are needed?

**Comment by flexiondotorg** [comment]
As discussed via email this can be closed. I'll file a new issue to request that the Vala Panel Appmenu registrar be separated into it's own service.

**Comment by rilian-la-te** [comment]
This will be opened for HUD itself, and registrar will be tracked in another issue, as we discussed.

**Comment by rilian-la-te** [comment]
Can you show me your HUD realization?

**Comment by flexiondotorg** [comment]
The `mate-hud` is in the GitHub repo:

  * https://github.com/ubuntu-mate/mate-hud/

When it is activated, it looks like this:

![mate-hud](https://user-images.githubusercontent.com/304639/29270995-365d7550-80f1-11e7-8246-5687eabe3034.gif)

You'll want to use release 17.10.2 because 17.10.3 requires a patch in Marco. The only difference between 17.10.2 and 17.10.3 is the keybinding to activate the HUD:

  * 1710.2: `Super` + `Alt`
  * 17.10.3: `Alt` *(only)*
